File: ABC100_199/ABC182/d.py
N = int(input())
A = list(map(int, input().split()))
# Sは累積和、Pは今の座標、Dはその時点で最大どこまで進むか
S = [0] * N
P = [0] * N
D = [0] * N

P[0] = A[0]
S[0] = A[0]
for i in range(1, N):
    S[i] += S[i - 1] + A[i]
    P[i] = P[i - 1] + S[i]

D = [0] * N
D[0] = S[0]
for i in range(N):
    D[i] = max(D[i - 1], S[i])

ans = max(A[0], 0)
for i in range(N - 1):
    ans = max(ans, P[i] + max(D[i], 0))
ans = max(ans, max(P))
#print(P, S, D)
print(ans)

File: ABC100_199/ABC182/d.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <vector>
using namespace std;
using ll = long long;

int main() {
  int N;
  cin >> N;
  vector<ll> A(N);
  for (int i = 0; i < N; i++) cin >> A[i];
 
  ll res = 0, ans = 0;
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < i + 1; j++) {
        res += A[j];
        if(A[j] > 0) ans = max(res, ans);
    }
  }
  cout << ans << endl;
  return 0;
}